### PR TITLE
issue9255 - Detect FIXME words in docstring

### LINKED ISSUE #9255: W0511: Doesn't detect TODO in docstrings
### Bug description

```python
# TODO: Write good doc strings.

def foobar():
    """
    TODO: Implement
    """
```

### Configuration

_No response_

### Command used

```shell
pylint todo.py
```

### Pylint output

```python
************* Module todo
todo.py:1:1: W0511: TODO: Write good doc strings. (fixme)
todo.py:1:0: C0114: Missing module docstring (missing-module-docstring)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

I expect the `TODO: Implement` to lead to an error reported by pylint.

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0]
```

### OS / Environment

Debian 12 in WSL 2

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by shenjunjie2003**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |
| ✓   | :scroll: Docs          |

## Description

Added checking in docstring. It checks for FIXME words (TODO, FIXME, XXX) in a docstring and give warnings.

Closes #9255

### CODE REVIEW

File: pylint/checkers/base/docstring_checker.py [hunk 1]
```diff
@@ -73,6 +73,11 @@ class DocStringChecker(_BasicChecker):
             "docstring.",
             {"old_names": [("C0111", "missing-docstring")]},
         ),
+        "C0150": (
+            "%s",
```
Comment (DanielNoord) [inline pylint/checkers/base/docstring_checker.py:77]: Could we make this message a little more descriptive? Something like:

```suggestion
            "Disallowed word in docstring: '%s'",
```

I know the original `fixme` message doesn't have this but I don't think that's nice :smile:

File: pylint/checkers/base/docstring_checker.py [hunk 2]
```diff
@@ -206,3 +211,14 @@ def _check_docstring(
             self.add_message(
                 "empty-docstring", node=node, args=(node_type,), confidence=confidence
             )
+        else:
+            fixme_words = self.linter.config.notes
+            for word in fixme_words:
```
Comment (DanielNoord) [inline pylint/checkers/base/docstring_checker.py:216]:
```suggestion
            for word in self.linter.config.notes:
```

File: tests/functional/d/docstrings_fixme.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# pylint: disable=unnecessary-pass, consider-using-f-string
+
+"""
+Test fixme in docstrings.
+"""
```
Comment (DanielNoord) [inline tests/functional/d/docstrings_fixme.py:5]: Could you switch the comment and the docstring around?

File: tests/functional/d/docstrings_fixme.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+# fixme-in-docstring
+def func2(): # [fixme-in-docstring]
+    """
+    FIXME: Implement
+    """
+
+# fixme-in-docstring
+def func3(): # [fixme-in-docstring]
+    """
+    XXX: Implement
```
Comment (DanielNoord) [inline tests/functional/d/docstrings_fixme.py:21]: Could you add a test where the `FIXME` is not the first word of the line? And also one where it is not surrounded by spaces?

**Review by DanielNoord — COMMENTED** [review]
This looks really good already!

You'll also need to add a documentation example. There is a CI check that sort of tells you what's wrong. Please let us know if you need more help figuring out what to do: that means we should update the error message of that CI step.

File: pylint/checkers/base/docstring_checker.py [hunk 5]
```diff
@@ -73,6 +73,11 @@ class DocStringChecker(_BasicChecker):
             "docstring.",
             {"old_names": [("C0111", "missing-docstring")]},
         ),
+        "C0150": (
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base/docstring_checker.py:76]: I would expect this message to be in the same checker than ``fixme``

File: pylint/checkers/base/docstring_checker.py [hunk 6]
```diff
@@ -73,6 +73,11 @@ class DocStringChecker(_BasicChecker):
...
+            "%s",
+            "fixme-in-docstring",
+            "Used when a warning note as FIXME or XXX is detected in docstring.",
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base/docstring_checker.py:79]: Imo should be a rational (why it's bad) and not when it's raised. (I know the current descriptions are a bad example of that.)

**Review by Pierre-Sassoulas — COMMENTED** [review]
If it's a new check, we might want to disable it by default. Not sure it warrant to create a new message by the way. If we extends ``fixme`` then adding an option to raise on docstring (false by default) would be nice.  What do you think @DanielNoord ?

**Comment by shenjunjie2003** [comment]
Ok! Should I change to integrate it to `fixme` ?

**Comment by Pierre-Sassoulas** [comment]
Yes, I think pylint would be more consistent if it's still a ``fixme`` message but with a ``check-fixme-in-docstring`` option (default to false). Here's an example of options in a checker: 
https://github.com/pylint-dev/pylint/blob/37081fdd40499d7d243f36849d39b2aa880c0e57/pylint/checkers/design_analysis.py#L292-L407

**Comment by DanielNoord** [comment]
@shenjunjie2003 are you still interested in working on this?

**Comment by shenjunjie2003** [comment]
Hi Daniel,

I'm sorry for not getting back to you sooner. I am no longer working on
this issue. Please feel free to find another developer.

Thanks

**Comment by badsketch** [comment]
Hey, I'd be interested in getting this over the finish line! Would be a new PR, but I'll cherrypick what I can from this branch so @shenjunjie2003 can get some credit

**Comment by DanielNoord** [comment]
Thanks @badsketch. I'll close this if your PR is open.

**Comment by Pierre-Sassoulas** [comment]
Closing in favor of #9744

**Pull request closed without merging**